doc/ir-semantics.text

```diff
@@ -121,49 +121,103 @@ declares i =
+drop all same-named bindings before a declaration.
+
+¹: the tightest possible assignment can be defined impredicatively as
+the pointwise intersection of all valid judgments, which is
+well-defined given that valid scopes are bounded by the set of
+variables defined in the program.
+
+  ∀(v → M) ∈ P,   A, P ⊨d v → M
+  ----------------------------
+            A, P ⊨d
+
+  ∀(l → i) ∈ M,   A, P ⊨d v, l → i
+  -------------------------------
+         A, P ⊨d v → M
+
+  S := A(v,l)
+  requires i ⊆ S
+  ∀l* ∈ pred(P(v),l), A(v,l*) ⊇ S
+  ∀l' ∈ succ(P(v),l), S ∪ declares i ⊇ A(v,l')
+  A, P ⊨d i : S
+  ----------------------------------------
+  A, P ⊨d v, l → i
+
+The judgment (A, P ⊢ i : S) below is the instruction-specific judgment
+that checks the scope restrictions that are specific to each
+instruction, rather than the general control-flow constraints and the
+generic requires/declares handling.
+
+  ----------------
+  A, P ⊨d start : ∅
+
+              A(v, l) = x*
+  ----------------------------------
+  A, P ⊨d osr(e, v, l, (x = ee*)) : S
+
+  ---------------
+  A, P ⊨d stop : ∅
+
+
+## Algorithmic scoping judgment
+
+(A, P ⊨a), (A, P ⊨a v → M), (A, P ⊨a i : S): as in the declarative
+scoping judgment. Only (A, P ⊨a v, l → i) differs.
+
+  S := A(v,l) \ declares i
+  (⋂_{l* ∈ pred(P(v),l)} A(v,l*)) = S ∪ requires i
+  A, P ⊨a i : S
+  ------------------------------------------------
+  A, P ⊨a v, l → i
+
+There is a fundamental difference in the role of the scope assignment
+A between the declarative and algorithmic judgments.
+
+In the declarative judgment, A(v,l) points to the variables
+*available* in the *ingoing* scope of the instruction at location
+(v,l): the scope that is available when arriving to the instruction
+from a predecessor. The variables that are available in the outgoing
+scope can be computed from this value as (A(v,l) ∪ declares i).
+
+In the algorithmic judgment, A(v,l) points to the variables that are
+*required* in the *outgoing* scope of the instruction at location
+(v,l): the scope that is required when leaving the instruction to
+a successor. The variables required in the ingoing scope can be
+computed from this value as (A(v,l) \ declares i).
+
+In the declarative judgment, the relation between an instruction scope
+and its neighbors is under-specified: a predecessor's outgoing scope
+may have more variables than our ingoing scope, and any successor's
+ingoing scope may have less variable than our outgoing scope.
+
+In the algorithmic judgment, the relation between the scopes of an
+instruction and its neighbors is deterministic: an instruction's
+ingoing scope is exactly the variables that are common among all its
+predecessors. Implicitly dropping a variable that will never be used
+again is possible: this is done by choosing a A(v,l) that is smaller
+than the instruction's ingoing set plus (declares i).
+
+Problem: I think there is a problem with the notion of outgoing scope
```

Comment:
The problem is with a small variable of the example of code you gave:

```
  { } mut a = 3
  {a} branch a x y
x:
  {a} print a
  stop
y:
  { } stop
```

it is compatible with the declarative judgment, but not with the algorithmic judgment: the scope of `y: { } stop` is wrong according to the rule that the scope here should be *exactly* the intersection of the scopes of all predecessors: the only predecessor here is `{a} branch`, so the scope must be `{a}` -- which means this judgment is inadequate to capture minimal scoping.

I think indeed that to fix this I would need to have, for each instruction, one scope per successor (so a set of annotations of the form `A(v,l,l')` in the algorithmic case). I think this starts to get very heavy and rather painful to reason about, which is what decided me to go the explicit `drop` route, where we would write instead:
```
  { } mut a = 3
  {a} branch a x y
x:
  {a} print a
  drop a
  stop
y:
  {a} drop a
  stop
```